File: PwrBot/FootballMatch.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Linq;
using System.Text.RegularExpressions;
using DotNetWikiBot;

namespace Simocracy.PwrBot
{
	[DebuggerDisplay("Home={HomeTeam}, Away={AwayTeam}, Result={Result}")]
	public class FootballMatch
	{
		/// <summary>
		/// Team, dessen Statistik erstellt wird, Basierend auf FLAGGENKÜRZEL
		/// </summary>
		public static string[] MainTeam { get; set; } = { "UNS" };

		private string _Tournament;
		public string Tournament
		{
			get { return _Tournament; }
			private set
			{
				var s = value.Trim();
				_Tournament = (s.Length < 3) ? s.Replace("-", String.Empty) : s;
			}
		}

		public DateTime Date { get; private set; }

		private string _City;
		/// <summary>
		/// Inkl. Flaggenkürzel
		/// </summary>
		public string City
		{
			get { return _City; }
			set { _City = value.Trim(); }
		}

		private string _Stadium;
		public string Stadium
		{
			get { return _Stadium; }
			private set { _Stadium = value.Trim(); }
		}

		private string _HomeTeam;
		/// <summary>
		/// Inkl. Flaggenkürzel
		/// </summary>
		public string HomeTeam
		{
			get { return _HomeTeam; }
			private set { _HomeTeam = value.Trim().Replace("'", ""); }
		}

		private string _AwayTeam;
		/// <summary>
		/// Inkl. Flaggenkürzel
		/// </summary>
		public string AwayTeam
		{
			get { return _AwayTeam; }
			private set { _AwayTeam = value.Trim().Replace("'", ""); }
		}

		/// <summary>
		/// Gegnerteam für Sortierung nach Gegner
		/// </summary>
		public string OpponentTeam
		{
			get
			{
				foreach(var s in MainTeam)
				{
					if(AwayTeam.Contains(s))
						return HomeTeam;
				}
				return AwayTeam;
			}
		}

		/// <summary>
		/// -1 wenn kein Spielergebnis
		/// </summary>
		public int ResultHome { get; private set; }

		/// <summary>
		/// -1 wenn kein Spielergebnis
		/// </summary>
		public int ResultAway { get; private set; }

		/// <summary>
		/// Vollständiges Spielergebnis
		/// </summary>
		public string Result
		{
			get
			{
				if(ResultHome < 0 || ResultAway < 0)
					return "X";
				else
					return ResultHome + ":" + ResultAway;
			}
		}

		public int Specators { get; private set; }
		public bool IsSOldOut { get; private set; }

		/// <summary>
		/// Inkl. Flaggenkürzel
		/// </summary>
		private string _Referee;
		public string Referee
		{
			get { return _Referee; }
			private set { _Referee = value.Trim(); }
		}

		/// <summary>
		/// Quellcode im Wiki
		/// </summary>
		public string SourceCode { get; private set; }

		public FootballMatch(Match match)
			: this(match.Groups[1].Value, match.Groups[2].Value, match.Groups[3].Value, match.Groups[4].Value, match.Groups[5].Value, match.Groups[6].Value, match.Groups[7].Value, match.Groups[8].Value, match.Groups[9].Value, match.Groups[10].Value, match.Value)
		{ }

		public FootballMatch(string tournament, string date, string city, string stadium, string homeTeam, string awayTeam, string result, string spectators, string soldOut, string referee, string source)
		{
			Tournament = tournament;
			SetDate(date);
			City = city;
			Stadium = stadium;
			HomeTeam = homeTeam;
			AwayTeam = awayTeam;
			SetResults(result);
			SetSpectators(spectators);
			SetIsSoldOut(soldOut);
			Referee = referee;
			SourceCode = source;
		}

		private void SetDate(string date)
		{
			DateTime dt = DateTime.MinValue;
			if(!String.IsNullOrEmpty(date))
			{
				var exactDatePattern = @"((\d{1,2})\.)?((\d{1,2})\.)?(\d{2,4})";
				var exactDateMatch = Regex.Match(date, exactDatePattern);
				if(exactDateMatch.Success)
				{
					var matchStr = exactDateMatch.Value;
					if(exactDateMatch.Groups[5].Length < 4)
						matchStr = matchStr.Insert(matchStr.Length - 2, "20");
					if(exactDateMatch.Groups[1].Success)
						DateTime.TryParse(matchStr, out dt);
					else
						dt = new DateTime(Int32.Parse(matchStr), 1, 1);
				}
			}
			Date = dt;
		}

		private void SetResults(string result)
		{
			var resPattern = @"(\d+):(\d+)";
			var resMatch = Regex.Match(result, resPattern);
			if(resMatch.Success)
			{
				ResultHome = Int32.Parse(resMatch.Groups[1].Value);
				ResultAway = Int32.Parse(resMatch.Groups[2].Value);
			}
			else
			{
				ResultHome = -1;
				ResultAway = -1;
			}
		}

		private void SetSpectators(string spectators)
		{
			int specs;
			Int32.TryParse(spectators, NumberStyles.Number, CultureInfo.CurrentCulture, out specs);
			Specators = specs;
		}

		private void SetIsSoldOut(string soldOut)
		{
			IsSOldOut = !String.IsNullOrEmpty(soldOut);
		}

		//public string GetLowLinesWikiCode()
		//{
		//	var regexMatch = Regex.Match(SourceCode, @"\|(.*[^\|\r\n]*)\s*[\r\n|\|]\|\s*([^\|\r\n]*)?\s*[\r\n|\|]\|\s*(\{\{.+\}\}[^\|\r\n]*)[\r\n\|]\|\s*([^\|\r\n]*)?\s*[\r\n\|]\|\s*('{0,3}(\{\{.+\}\})[^\|\r\n]*)?\s*[\r\n\|]\|\s*('{0,3}(\{\{.+\}\})[^\|\r\n]*)?\s*[\r\n\|]\|\s*[^\|]*\|\s*'''((\d+):(\d+))?([^']*)'''(\s*(<br>)?\s*([^\|\r\n]*))?\s*[\r\n\|]\|\s*([\d\.]*)(\s*\(a\))?\s*[\r\n\|]\|\s*(\{\{.*\}\}[^\|\r\n]*)?");
		//	var tournament = regexMatch.Groups[1].Value.Trim();
		//	var date = regexMatch.Groups[2].Value.Trim();
		//	var city = regexMatch.Groups[3].Value.Trim();
		//	var stadium = regexMatch.Groups[4].Value.Trim();
		//	var referee = regexMatch.Groups[18].Value.Trim();
			
		//	var specs = regexMatch.Groups[16].Value.Trim();
		//	if(!String.IsNullOrEmpty(regexMatch.Groups[17].Value))
		//		specs += " " + regexMatch.Groups[17].Value.Trim();

		//	int resultHome = -1;
		//	int resultAway = -1;
		//	var hasRes = Int32.TryParse(regexMatch.Groups[10].Value, out resultHome);
		//	if(hasRes)
		//		hasRes = Int32.TryParse(regexMatch.Groups[11].Value, out resultAway);
		//	var ResComment = regexMatch.Groups[12].Value.Trim();
		//	var halfRes = regexMatch.Groups[13].Value.TrimEnd();
		//	string result = "'''";
		//	if(hasRes)
		//	{
		//		result += resultHome + ":" + resultAway;
		//		if(!String.IsNullOrEmpty(ResComment))
		//			result += " " + ResComment;
		//		result += "'''" + halfRes;
		//	}
		//	else
		//	{
		//		result += ResComment + "'''";
		//	}

		//	string homeTeam = regexMatch.Groups[6].Value.Replace("'", "").Trim();
		//	if(homeTeam.Contains("?"))
		//	{
		//		homeTeam = regexMatch.Groups[5].Value.Replace("'", "").Trim();
		//	}
		//	else if(!homeTeam.Contains("#"))
		//	{
		//		homeTeam = homeTeam.Insert(homeTeam.Length - 2, "|#");
		//	}
		//	if(resultHome > resultAway)
		//		homeTeam = "'''" + homeTeam + "'''";

		//	string awayTeam = regexMatch.Groups[8].Value.Replace("'", "").Trim();
		//	if(awayTeam.Contains("?"))
		//	{
		//		awayTeam = regexMatch.Groups[7].Value.Replace("'", "").Trim();
		//	}
		//	else if(!awayTeam.Contains("#"))
		//	{
		//		awayTeam = awayTeam.Insert(awayTeam.Length - 2, "|#");
		//	}
		//	if(resultAway > resultHome)
		//		awayTeam = "'''" + awayTeam + "'''";

		//	return String.Format("| {0}\n| {1} || {2} || {3}\n| {4} || {5}\n|style=\"text-align:center;\"| {6}\n| {7} || {8}\n",
		//		tournament, // Turnier
		//		date, // Datum
		//		city, // Ort
		//		stadium, // Stadion
		//		homeTeam, // Heim
		//		awayTeam, // Aus
		//		result, // Ergebnis
		//		specs, // Zuschauer
		//		referee); // Schiri
		//}

		/// <summary>
		/// Parsed die Matches in <see cref="FootballMatch"/>-Objekte
		/// </summary>
		/// <param name="matches">Matches</param>
		/// <returns>Liste mit <see cref="FootballMatch"/></returns>
		public static List<FootballMatch> GetMatchList(MatchCollection matches)
		{
			var matchesList = new List<FootballMatch>(matches.Count);
			foreach(Match match in matches)
				matchesList.Add(new FootballMatch(match));
			return matchesList;
		}

		/// <summary>
		/// Analysiert die Statistik und gibt diese nach Gegnern gruppiert und alphabetisch sortiert zurück
		/// </summary>
		/// <param name="matches">Spiele</param>
		/// <returns>Statistik</returns>
		public static SortedDictionary<string, FootballStatElement> GroupByOpponents(IEnumerable<FootballMatch> matches)
		{
			var sdic = new SortedDictionary<string, FootballStatElement>();
			var flagTempRegex = new Regex(@"\{\{([^\|\}]*)(\|([^\}\|]*))?(\|([^\}\|]*))?\}\}");

			foreach(var match in matches)
			{
				var flagMatch = flagTempRegex.Match(match.OpponentTeam);

				var flag = flagMatch.Groups[1].Value;
				if(String.IsNullOrEmpty(flag))
					continue;

				int pxValue;
				string name;
				// {{GRA|Grafenberg}} -> Grafenberg
				if(flagMatch.Groups[3].Success && !flagMatch.Groups[3].Value.Contains("=") && !Int32.TryParse(flagMatch.Groups[3].Value, out pxValue))
				{
					name = flagMatch.Groups[3].Value;
				}
				// {{GRA|b=20|Grafenberg}} -> Grafenberg
				else if(flagMatch.Groups[5].Success && !flagMatch.Groups[5].Value.Contains("=") && !Int32.TryParse(flagMatch.Groups[5].Value, out pxValue))
				{
					name = flagMatch.Groups[5].Value;
				}
				// {{GRA}} -> "" -> {{GRA}} Grafenberg -> Grafenberg
				else
				{
					name = flagTempRegex.Replace(match.OpponentTeam, String.Empty).Trim();
				}

				var isNoneFlag = flag.Contains("?");
				var isDummyName = name.Contains("#");
				flag = Simocracy.GetFlag(isNoneFlag ? name : flag);
				name = Simocracy.GetStateName(isDummyName ? flag : name);

				if(!sdic.ContainsKey(name))
				{
					var fse = new FootballStatElement()
					{
						Name = name,
						Flag = $"{{{{{flag}|#}}}}"
					};
					sdic.Add(name, fse);
				}
				sdic[name].AddMatch(match);
			}

			return sdic;
		}

		/// <summary>
		/// Ausgabe der Bilanz von <see cref="MainTeam"/> nach Gegnern
		/// </summary>
		/// <param name="opponents">Bilanz als <see cref="SortedDictionary{String, FootballStatElement}"/></param>
		/// <returns>Bilanz als Wikicode</returns>
		public static string GetOpponentTableCode(SortedDictionary<string, FootballStatElement> opponents)
		{
			var text = "=== Nach Gegner ===\n" +
				"<html>"/*"<style>\n" +
				".s {\n" +
				"    background:#CCFFCC;\n" +
				"}\n" +
				".u {\n" +
				"    background:#FFFFCC;\n" +
				"}\n" +
				".n {\n" +
				"    background:#FFCCCC;\n" +
				"}</style></html>\n"*/ +
				"{| class=\"wikitable sortable\" style=\"text-align:center;\"\n" +
				"|-\n" +
				"! Mannschaft\n" +
				"! <abbr title=\"Spiele\">Sp.</abbr>\n" +
				"! <abbr title=\"Siege\">S</abbr>\n" +
				"! <abbr title=\"Unentschieden\">U</abbr>\n" +
				"! <abbr title=\"Niederlagen\">N</abbr>\n" +
				"! <abbr title=\"Tore\">T</abbr>\n" +
				"! <abbr title=\"Gegentore\">GT</abbr>\n" +
				"! <abbr title=\"Tordifferenz\">TD</abbr>\n" +
				"! <abbr title=\"Punkte\">P</abbr>";

			foreach(var opp in opponents)
			{
				if(!MainTeam.Contains(opp.Value.Flag))
					text = $"{text}\n{opp.Value.OpponentWikicode}";
			}
			
			text = $"{text}\n|}}\n<sup>Stand: <drechner eing=\"j\" day=\"j\">{DateTime.Now:yyyy-MM-dd HH:mm}</drechner></sup>";

			return text;
		}

		public static SortedDictionary<int, FootballStatElement> GroupByYears(IEnumerable<FootballMatch> matches)
		{
			var sdic = new SortedDictionary<int, FootballStatElement>();

			foreach(var match in matches)
			{
				if(match.Date != DateTime.MinValue && !MainTeam.Contains(Regex.Match(match.OpponentTeam, @"\{\{([^\|\}]*)(\|([^\}\|]*))?(\|([^\}\|]*))?\}\}").Groups[1].Value))
				{
					if(!sdic.ContainsKey(match.Date.Year))
					{
						var fse = new FootballStatElement()
						{
							Year = match.Date.Year
						};
						sdic.Add(match.Date.Year, fse);
					}
					sdic[match.Date.Year].AddMatch(match);
				}
			}

			return sdic;
		}

		public static string GetYearTableCode(SortedDictionary<int, FootballStatElement> years)
		{
			var text = "=== Nach Jahr ===\n" +
				"{| class=\"wikitable sortable\" style=\"text-align:center;\"\n" +
				"|-\n" +
				"! Jahr\n" +
				"! <abbr title=\"Spiele\">Sp.</abbr>\n" +
				"! <abbr title=\"Siege\">S</abbr>\n" +
				"! <abbr title=\"Unentschieden\">U</abbr>\n" +
				"! <abbr title=\"Niederlagen\">N</abbr>\n" +
				"! <abbr title=\"Tore\">T</abbr>\n" +
				"! <abbr title=\"Gegentore\">GT</abbr>\n" +
				"! <abbr title=\"Tordifferenz\">TD</abbr>\n" +
				"! <abbr title=\"Punkte\">P</abbr>";

			int allWon = 0;
			int allDrawn = 0;
			int allLose = 0;
			int allGoalsFor = 0;
			int allGoalsAgainst = 0;

			foreach(var year in years)
			{
				allWon += year.Value.Win;
				allDrawn += year.Value.Drawn;
				allLose += year.Value.Lose;
				allGoalsFor += year.Value.GoalsFor;
				allGoalsAgainst += year.Value.GoalsAgainst;

				text = $"{text}\n{year.Value.YearWikicode}";
			}

			var played = allWon + allDrawn + allLose;
			var allGoalDiff = allGoalsFor - allGoalsAgainst;
			var points = allWon * 3 + allDrawn;
			text =
				$"{text}\n|-\n! Ges. || {played} || {allWon} || {allDrawn} || {allLose} || {allGoalsFor} || {allGoalsAgainst} || {allGoalDiff:+0;-0;+0} || {points}\n|}}" +
				$"\n<sup>Stand: <drechner eing=\"j\" day=\"j\">{DateTime.Now:yyyy-MM-dd HH:mm}</drechner></sup>";

			return text;
		}

		/// <summary>
		/// Analysiert und wertet die Statistik einer Fußballnationalmannschaft aus
		/// </summary>
		/// <param name="articleName">Artikel mit der Statistik</param>
		/// <param name="mainTeams">Mainteams, auf die ausgewertet werden soll</param>
		public static void AnalyseFootballStats(string articleName, params string[] mainTeams)
		{
			MainTeam = mainTeams;
			/*
			 * Gruppen:
			 * 1: Turnier
			 * 2: Datum
			 * 3: Ort
			 * 4: Stadion
			 * 5: Heimteam
			 * 6: Gastteam
			 * 7: Endergebnis (120 min)
			 * 8: Zuschauer
			 * 9: Wenn nicht leer: Ausverkauft
			 * 10: Schiedsrichter
			 */
			var matchRegexPat = @"\|(.*[^\|\r\n]*)\s*[\r\n|\|]\|\s*([^\|\r\n]*)?\s*[\r\n|\|]\|\s*(\{\{.+\}\}[^\|\r\n]*)[\r\n\|]\|\s*([^\|\r\n]*)?\s*[\r\n\|]\|\s*('{0,3}\{\{.+\}\}[^\|\r\n]*)?\s*[\r\n\|]\|\s*('{0,3}\{\{.+\}\}[^\|\r\n]*)?\s*[\r\n\|]\|\s*[^\|]*\|\s*'''([^']*)'''[^\|\r\n]*?\s*[\r\n\|]\|\s*([\d\.]*)(\s*\(a\))?\s*[\r\n\|]\|\s*(\{\{.*\}\}[^\|\r\n]*)?";

			Console.WriteLine("Lade Artikel");
			Page p = new Page(articleName);
			p.Load();

			var sectionMatches = Regex.Matches(p.text, @"(^={1,6}(.*)?={1,6}\s*?$)", RegexOptions.Multiline);
			int opponentSection = 0;
			int yearSection = 0;
			for(int i = 0; i < sectionMatches.Count; i++)
			{
				if(sectionMatches[i].Value.Contains("Nach Gegner"))
					opponentSection = i + 1;
				else if(sectionMatches[i].Value.Contains("Nach Jahr"))
					yearSection = i + 1;
			}

			var matchesRegex = Regex.Matches(p.text, matchRegexPat);
			var footballMatchList = GetMatchList(matchesRegex);
			Console.WriteLine(footballMatchList.Count + " Matches found");

			//foreach(var match in footballMatchList)
			//{
			//	var newText = match.GetLowLinesWikiCode();
			//	p.text = p.text.Replace(match.SourceCode, newText);
			//}

			//p.Save();

			if(opponentSection != 0)
			{
				var opponents = GroupByOpponents(footballMatchList);
				Console.WriteLine(opponents.Count + " Opponents found");
				foreach(var o in opponents)
					Console.WriteLine("{0}: {1} Matches", o.Key, o.Value.Played);

				var sectionText = GetOpponentTableCode(opponents);

				var editStr = String.Format("action=edit&format=xml&bot=1&title={0}&section={1}&text={2}&summary={3}&token={4}",
					Bot.UrlEncode(articleName),
					Bot.UrlEncode(opponentSection.ToString()),
					Bot.UrlEncode(sectionText),
					Bot.UrlEncode("Aktualisierung der Gegnerstatistiken"),
					Bot.UrlEncode(PwrBot.Site.tokens["csrftoken"]));
				var result = PwrBot.Site.PostDataAndGetResult(PwrBot.Site.apiPath, editStr);
				Console.WriteLine("Opponent Result:");
				Console.WriteLine(result);
			}

			if(yearSection != 0)
			{
				var years = FootballMatch.GroupByYears(footballMatchList);
				Console.WriteLine(years.Count + " Years found");
				foreach(var o in years)
					Console.WriteLine("{0}: {1} Matches", o.Key, o.Value.Played);

				var sectionText = FootballMatch.GetYearTableCode(years);

				var editStr = String.Format("action=edit&format=xml&bot=1&title={0}&section={1}&text={2}&summary={3}&token={4}",
					Bot.UrlEncode(articleName),
					Bot.UrlEncode(yearSection.ToString()),
					Bot.UrlEncode(sectionText),
					Bot.UrlEncode("Aktualisierung der Jahresstatistiken"),
					Bot.UrlEncode(PwrBot.Site.tokens["csrftoken"]));
				var result = PwrBot.Site.PostDataAndGetResult(PwrBot.Site.apiPath, editStr);
				Console.WriteLine("Year Result:");
				Console.WriteLine(result);
			}
		}
	}
}

File: WikiAnalyser/footballMatchAnalyzer/footballMatch.py
import re
from datetime import datetime
from collections import OrderedDict

from simocracy import flagConverter
from simocracy import wiki
from footballMatchAnalyzer.footballStatElement import FootballStatElement

class FootballMatch:
	"""
	Fußballspiel
	"""

	"""
	Team, dessen Statistik erstellt wird, Basierend auf FLAGGENKÜRZEL
	"""
	mainTeam = ["UNS"]

	# Instance Fields

	_tournament = ""
	@property
	def tournament(self):
		return self._tournament
	@tournament.setter
	def tournament(self, value):
		s = value.strip()
		self._Tournament = s.replace("-", "") if (len(s) < 3) else s

	date = datetime.min

	_city = ""
	@property
	def city(self):
		return self._city
	@city.setter
	def city(self, value):
		self._city = value.strip()

	_stadium = ""
	@property
	def stadium(self):
		return self._stadium
	@stadium.setter
	def stadium(self, value):
		self._stadium = value.strip()

	_homeTeam = ""
	@property
	def homeTeam(self):
		"""
		Inkl. Flaggenkürzel
		"""
		return self._homeTeam
	@homeTeam.setter
	def homeTeam(self, value):
		self._homeTeam = value.strip().replace("'", "")

	_awayTeam = ""
	@property
	def awayTeam(self):
		"""
		Inkl. Flaggenkürzel
		"""
		return self._awayTeam
	@awayTeam.setter
	def awayTeam(self, value):
		self._awayTeam = value.strip().replace("'", "")
		
	@property
	def opponentTeam(self):
		"""
		Gegnerteam für Sortierung nach Gegner
		"""
		for s in FootballMatch.mainTeam:
			if s in self.awayTeam:
				return self.homeTeam
		return self.awayTeam
		
	"""
	-1 wenn kein Spielergebnis
	"""
	resultHome = ""

	"""
	-1 wenn kein Spielergebnis
	"""
	resultAway = ""

	@property
	def result(self):
		"""
		Vollständiges Spielergebnis
		"""
		if (self.resultHome < 0 or self.resultAway < 0):
			return "X"
		else:
			return str(self.resultHome) + ":" + str(self.resultAway)

	spectators = 0
	isSoldOut = False

	_referee = ""
	@property
	def referee(self):
		"""
		Inkl. Flaggenkürzel
		"""
		return self._referee
	@referee.setter
	def referee(self, value):
		self._referee = value.strip().replace("'", "")

	def __init__(self, *args, **kwargs):
		if len(args) == 10:
			self.initFromInput(args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7], args[8], args[9])
		elif len(args) == 1:
			self.initFromMatch(args[0])
		else:
			return None

	def initFromMatch(self, reMatch):
		self.initFromInput(reMatch[0], reMatch[1], reMatch[2], reMatch[3], reMatch[4], reMatch[5], reMatch[6], reMatch[7], reMatch[8], reMatch[9])

	def initFromInput(self, tournament, date, city, stadium, homeTeam, awayTeam, result, spectators, soldOut, referee):
			self.tournament = tournament;
			self.setDate(date);
			self.city = city;
			self.stadium = stadium;
			self.homeTeam = homeTeam;
			self.awayTeam = awayTeam;
			self.setResults(result);
			self.setSpectators(spectators);
			self.SetIsSoldOut(soldOut);
			self.referee = referee;

	def setDate(self, date):
		try:
			exactDatePattern = r"((\d{1,2})\.)?((\d{1,2})\.)?(\d{2,4})"
			exactDateMatch = re.match(exactDatePattern, date)
			if not exactDateMatch is None:
				matchStr = exactDateMatch.group(0)
				if len(exactDateMatch.group(5)) < 4:
					matchStr = matchStr[:len(matchStr)-2] + "20" + matchStr[len(matchStr)-2:]
				if exactDateMatch.group(2) is None:
					datepstr = "%Y"
				elif exactDateMatch.group(4) is None:
					datepstr = "%m.%Y"
				else:
					datepstr = "%d.%m.%Y"
				parsedDate = datetime.strptime(matchStr, datepstr)
				self.date = parsedDate

		except:
			self.date = datetime.min
		

	def setResults(self, result):
		resPattern = r"(\d+):(\d+)"
		resMatch = re.match(resPattern, result)
		if not resMatch is None:
			self.resultHome = int(resMatch.group(1))
			self.resultAway = int(resMatch.group(2))
		else:
			self.resultHome = -1;
			self.resultAway = -1;

	def setSpectators(self, spectators):
		try:
			sp = spectators.replace(".","").replace("'","").replace(" ","")
			self.spectators = int(sp)
		except:
			self.spectators = 0

	def SetIsSoldOut(self, soldOut):
		self.isSoldOut = len(soldOut.strip()) > 0

def getMatchList(matches):
	"""
	Parsed die Matches in FootballMatch-Objekte
	matches (Match-Collection): Matches
	return: Liste mit FootballMatch
	"""
	matchList = []
	for match in matches:
		matchList.append(FootballMatch(match))
	return matchList
	
def groupByOpponents(matches):
	"""
	Analysiert die Statistik und gibt diese nach Gegnern gruppiert und alphabetisch sortiert zurück
	matches (Enumerable<FootballMatch>): Spiele
	return (OrderedDict<string, FootballStatElement>): Statistik
	"""
	dic = {}
	flagTempRegex = re.compile(r"\{\{([^\|\}]*)(\|([^\}\|]*))?(\|([^\}\|]*))?\}\}")
	for match in matches:
		flagMatch = flagTempRegex.match(match.opponentTeam)

		if flagMatch is None:
			continue
		flag = flagMatch.group(1)

		# {{GRA|Grafenberg}} -> Grafenberg
		if not flagMatch.group(3) is None and not "=" in flagMatch.group(3) and re.match(r"\d+", flagMatch.group(3)) is None:
			name = flagMatch.group(3)
		# {{GRA|b=20|Grafenberg}} -> Grafenberg
		elif not flagMatch.group(5) is None and not "=" in flagMatch.group(5) and re.match(r"\d+", flagMatch.group(5)) is None:
			name = flagMatch.group(5)
		# {{GRA}} -> "" -> {{GRA}} Grafenberg -> Grafenberg
		else:
			name = flagTempRegex.sub("", match.opponentTeam).strip()

		isNoneFlag = "?" in flag
		isDummyName = "#" in name
		flag = flagConverter.getFlag(name if isNoneFlag else flag)
		name = flagConverter.getStateName(flag if isDummyName else name)

		if not name in dic:
			fse = FootballStatElement(name, flag)
			dic[name] = fse
		dic[name].addMatch(match)

	sdic = OrderedDict(sorted(dic.items()))
	return sdic

def getOpponentTableCode(opponents):
	"""
	Ausgabe der Bilanz von MainTeam nach Gegnern
	opponents: Bilanz als OrderedDict<string, FootballStatElement>
	return: Bilanz als Wikicode
	"""
	text = ("=== Nach Gegner ===\n"
			"{| class=\"wikitable sortable\" style=\"text-align:center;\"\n"
			"|-\n"
			"! Mannschaft\n"
			"! <abbr title=\"Spiele\">Sp.</abbr>\n"
			"! <abbr title=\"Siege\">S</abbr>\n"
			"! <abbr title=\"Unentschieden\">U</abbr>\n"
			"! <abbr title=\"Niederlagen\">N</abbr>\n"
			"! <abbr title=\"Tore\">T</abbr>\n"
			"! <abbr title=\"Gegentore\">GT</abbr>\n"
			"! <abbr title=\"Tordifferenz\">TD</abbr>\n"
			"! <abbr title=\"Punkte\">P</abbr>")
	for opp in opponents.items():
		flag = re.match(r"\{\{([^\|\}]*)(\|([^\}\|]*))?(\|([^\}\|]*))?\}\}", opp[1].flag)
		if not flag is None and not flag.group(1) in FootballMatch.mainTeam:
			text = str.format("{0}\n{1}", text, opp[1].opponentWikicode)

	text = str.format("{0}\n|}}\n<sup>Stand: <drechner eing=\"j\" day=\"j\">{1:%Y-%m-%d %H:%M}</drechner></sup>", text, datetime.now())

	return text

def groupByYears(matches):
	dic = {}
	for match in matches:
		flag = re.match(r"\{\{([^\|\}]*)(\|([^\}\|]*))?(\|([^\}\|]*))?\}\}", match.opponentTeam)
		if not match.date is datetime.min and not flag is None and not flag.group(1) in FootballMatch.mainTeam:
			if not match.date.year in dic:
				fse = FootballStatElement(match.date.year)
				dic[match.date.year] = fse
			dic[match.date.year].addMatch(match)

	sdic = OrderedDict(sorted(dic.items()))
	return sdic

def getYearTableCode(years):
	text = ("=== Nach Jahr ===\n"
			"{| class=\"wikitable sortable\" style=\"text-align:center;\"\n"
			"|-\n"
			"! Jahr\n"
			"! <abbr title=\"Spiele\">Sp.</abbr>\n"
			"! <abbr title=\"Siege\">S</abbr>\n"
			"! <abbr title=\"Unentschieden\">U</abbr>\n"
			"! <abbr title=\"Niederlagen\">N</abbr>\n"
			"! <abbr title=\"Tore\">T</abbr>\n"
			"! <abbr title=\"Gegentore\">GT</abbr>\n"
			"! <abbr title=\"Tordifferenz\">TD</abbr>\n"
			"! <abbr title=\"Punkte\">P</abbr>")

	allWon = 0;
	allDrawn = 0;
	allLose = 0;
	allGoalsFor = 0;
	allGoalsAgainst = 0;

	for year in years.items():
		allWon += year[1].win
		allDrawn += year[1].drawn
		allLose += year[1].lose
		allGoalsFor += year[1].goalsFor
		allGoalsAgainst += year[1].goalsAgainst

		text = str.format("{0}\n{1}", text, year[1].yearWikicode)

	played = allWon + allDrawn + allLose
	allGoalDiff = allGoalsFor - allGoalsAgainst
	points = allWon * 3 + allDrawn

	text = str.format("{0}\n|-\n! Ges. || {1} || {2} || {3} || {4} || {5} || {6} || {7:+0d} || {8}\n|}}" +
				"\n<sup>Stand: <drechner eing=\"j\" day=\"j\">{9:%Y-%m-%d %H:%M}</drechner></sup>",
			text, played, allWon, allDrawn, allLose, allGoalsFor, allGoalsAgainst, allGoalDiff, points, datetime.now())

	return text

def analyseFootballStats(articleName, mainTeams):
	"""
	Analysiert und wertet die Statistik einer Fußballnationalmannschaft aus
	articleName: Artikel mit der Statistik
	mainTeams[]: Mainteams, auf die ausgewertet werden soll
	"""
	FootballMatch.mainTeam = mainTeams

	"""
	Gruppen:
	1: Turnier
	2: Datum
	3: Ort
	4: Stadion
	5: Heimteam
	6: Gastteam
	7: Endergebnis (120 min)
	8: Zuschauer
	9: Wenn nicht leer: Ausverkauft
	10: Schiedsrichter
	"""
	matchRegexPat = r"\|(.*[^\|\r\n]*)\s*[\r\n|\|]\|\s*([^\|\r\n]*)?\s*[\r\n|\|]\|\s*(\{\{.+\}\}[^\|\r\n]*)[\r\n\|]\|\s*([^\|\r\n]*)?\s*[\r\n\|]\|\s*('{0,3}\{\{.+\}\}[^\|\r\n]*)?\s*[\r\n\|]\|\s*('{0,3}\{\{.+\}\}[^\|\r\n]*)?\s*[\r\n\|]\|\s*[^\|]*\|\s*'''([^']*)'''[^\|\r\n]*?\s*[\r\n\|]\|\s*([\d\.]*)(\s*\(a\))?\s*[\r\n\|]\|\s*(\{\{.*\}\}[^\|\r\n]*)?"

	print("Lade Artikel " + articleName)
	p = wiki.Article(articleName)
		
	# Abschnitte analysieren
	sectionMatches = re.findall(r"(^={1,6}(.*)?={1,6}\s*?$)", p.asString, re.MULTILINE)
	
	counter = 0
	opponentSection = 0
	yearSection = 0
	for sectMatch in sectionMatches:
		counter += 1
		if "Gegner" in sectMatch[0]:
			opponentSection = counter
		elif "Jahr" in sectMatch[0]:
			yearSection = counter

	matchesRegex = re.findall(matchRegexPat, p.asString)
	footballMatchList = getMatchList(matchesRegex)
	print(str(len(footballMatchList)) + " Spiele gefunden")

	# Gegnerstats
	if opponentSection > 0:
		opponents = groupByOpponents(footballMatchList)
		print(str(len(opponents)) + " Gegner gefunden")
		for o in opponents.items():
			print(o[0] + ": " + str(o[1].played) + " Spiele")

		sectionText = getOpponentTableCode(opponents)
		wiki.edit_article(articleName, sectionText, opponentSection)

	# Jahresstats
	if yearSection > 0:
		years = groupByYears(footballMatchList)
		print(str(len(years)) + " Jahre gefunden")
		for y in years.items():
			print(str(y[0]) + ": " + str(y[1].played) + " Spiele")

		sectionText = getYearTableCode(years)
		wiki.edit_article(articleName, sectionText, yearSection)
